src/labthings_fastapi/thing_property.py

```diff
@@ -0,0 +1,988 @@
+    # type of ``self`` when decorating methods, so it is left as
+    # ``Any`` to avoid the major confusion that would result from
+    # trying to type it more tightly.
+    #
+    # Note: in ``@overload`` definitions, it's sometimes necessary
+    # to avoid the use of these aliases, as ``mypy`` can't
+    # pick which variant is in use without the explicit `Callable`.
+    ValueFactory: TypeAlias = Callable[[], Value]
+    ValueGetter: TypeAlias = Callable[[Any], Value]
+    ValueSetter: TypeAlias = Callable[[Any, Value], None]
+
+
+def default_factory_from_arguments(
+    default: Value | EllipsisType = ...,
+    default_factory: ValueFactory | None = None,
+) -> ValueFactory:
+    """Process default arguments to get a default factory function.
+
+    This function takes the ``default`` and ``default_factory`` arguments
+    and will either return the ``default_factory`` if it is provided, or
+    will wrap the default value provided in a factory function.
+
+    This is used to avoid repeating the logic of checking whether a default
+    value or a factory function has been provided, and it returns a factory
+    rather than a default value so that it may be called multiple times to
+    get copies of the default value.
+
+    This function also ensures the default is specified exactly once, and
+    raises exceptions if it is not.
+
+    This logic originally lived only in the initialiser of `.DataProperty`
+    but it was needed in the `.property` and `.setting` functions in order
+    to correctly type them (so that specifying both or neither of the
+    ``default`` and ``default_factory`` arguments would raise an error
+    with mypy).
+
+    :param default: the default value, or an ellipsis if not specified.
+    :param default_factory: a function that returns the default value.
+    :return: a function that returns the default value.
+    :raises OverspecifiedDefaultError: if both ``default`` and
+        ``default_factory`` are specified.
+    :raises MissingDefaultError: if neither ``default`` nor ``default_factory``
+        are specified.
+    """
+    if default is ... and default_factory is None:
+        # If the default is an ellipsis, we have no default value.
+        # This is not allowed for DataProperty, so we raise an error.
+        raise MissingDefaultError()
+    if default is not ... and default_factory is not None:
+        # If both default and default_factory are set, we raise an error.
+        raise OverspecifiedDefaultError()
+    if default is not ...:
+
+        def default_factory() -> Value:
+            return default
+
+    assert callable(default_factory), "The default_factory must be callable."
+    return default_factory
+
+
+# See comment at the top of the file regarding ignored linter rules.
+@overload  # use as a decorator  @property
+def property(  # noqa: D103
+    getter: Callable[[Any], Value],
+) -> FunctionalProperty[Value]: ...
+@overload  # use as `field: int = property(0)``
```

Comment:
Is this true? Isn't default needed?
```suggestion
@overload  # use as `field: int = property(default=0)``
```

I am not overly familiar with overload. Also what happens if default is not specified, does it try to set the getter to 0 and then error?

I very much agree this code could do with spacing. I have had to check it out locally to read it.

Reply:
Yes, if it's called with a single positional argument that is not callable, it will fail with an error. I'll add a test for that.